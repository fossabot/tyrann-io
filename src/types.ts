import * as t from 'io-ts'
import * as E from 'io-ts/Encoder'

export type BaseMethods = "get" | "post" | "put" | "delete" | "options" | "patch";

export type TyrannApis = {
  [path in string]: Path;
}

export interface Path {
  get?: Operation;
  post?: Operation;
  put?: Operation;
  delete?: Operation;
  trace?: Operation;
  options?: Operation;
  patch?: Operation;
}

export interface Operation {
  query?: t.Any;
  path?: t.Any;
  body?: t.Any;
  response: {
    [code in number]: t.Any;
  },
  options?: OperationOptions;
}

export interface OperationOptions {
  queryEncoder?: E.Encoder<string, any>;
  pathEncoder?: E.Encoder<string, any>;
  bodyEncoder?: E.Encoder<string, any>;
}

export interface Request<
  Query extends {} = any,
  Path extends {} = any,
  Body extends {} = any,
  > {
  query?: Query;
  path?: Path;
  body?: Body;
}

export type MethodsOf<P extends Path> = keyof P & BaseMethods;

export type RequestOf<O extends Operation> = (
  O['query'] extends t.Any ? { query: t.TypeOf<O['query']>  } : {}
) & (
  O['path'] extends t.Any ? { path: t.TypeOf<O['path']> } : {}
) & (
  O['body'] extends t.Any  ? { body: t.TypeOf<O['body']> } : {}
);

export type ResponseOf<O extends Operation> = {
  [code in keyof O['response'] & number]?: t.TypeOf<O['response'][code]>
};
