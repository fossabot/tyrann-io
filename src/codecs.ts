import queryString, { ParseOptions } from 'query-string';
import format from 'string-format';
import * as E from 'io-ts/Encoder';


export const createQueryEncoder = <O extends {}>(
  opts?: ParseOptions,
): E.Encoder<string, O> => {
  return {
    encode: (o: O) => queryString.stringify(o, opts)
  };
}

export type QueryEncoder = ReturnType<typeof createQueryEncoder>;


export const createJsonEncoder = <O extends {}>(
  ...jsonArgs: any[]
): E.Encoder<string, O> => {
  return {
    encode: (o: O) => JSON.stringify(o, ...jsonArgs)
  };
}

export type JsonEncoder = ReturnType<typeof createJsonEncoder>;

// Decoders seems not to be interoperable with t.Type, give up using it for the time.

// export const jsonDecoder = <T extends t.Any>(
//   type: T,
// ): D.Decoder<string, t.TypeOf<T>> => {
//   return {
//     decode: (s: string) => {
//       const result = JSON.parse(s);
//       return type.asDecoder().decode(result);
//     },
//   }
// }

export const createPathEncoder = <O extends {}>(
  path: string,
): E.Encoder<string, O> => {
  return {
    encode: (o: O) => format(path, o),
  };
}

export const createIdentityEncoder = <O extends {}>(): E.Encoder<O, O> => {
  return {
    encode: (o: O) => o,
  };
}

export type IdentityEncoder = ReturnType<typeof createIdentityEncoder>;