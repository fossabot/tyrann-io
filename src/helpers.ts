import { Either, isLeft, isRight } from 'fp-ts/lib/Either';
import * as t from 'io-ts';

export type OmittableKeys<A extends {}> = {
  [K in keyof A]: undefined extends A[K] ? K : null extends A[K] ? K : undefined
}[keyof A] & string;


export type Defaultable<A extends {}> = {
  [K in keyof A & OmittableKeys<A>]?:  A[K]
} & {
  [K in Exclude<keyof A, OmittableKeys<A>>]:  A[K]
}

export const transformable = <AType extends t.Any, OutType, A = t.TypeOf<AType>>(
  a: AType,
  transform: (a: A) => OutType,
) => {
  return new t.Type<A, OutType>(
    `Transformable<${a.name}>`,
    a.is,
    a.validate,
    transform,
  );
};

export const defaultable = <AType extends t.Any, Placeholder, A = t.TypeOf<AType>>(
  a: AType,
  placeholder: Placeholder,
) => {
  return transformable<t.Type<Defaultable<A>>, Defaultable<A> & Placeholder>(
    a,
    (a: Defaultable<A>) => ({
      ...placeholder,
      ...a,
    })
  );
}

export const omittable = <AType extends t.Any>(
  a: AType,
) => {
  return t.union([
    a,
    t.nullType,
    t.undefined,
  ]);
}

export type Validate<A, I> = (input: I, context: t.Context) => Either<t.Errors, A>; 
export interface Validator<A, I> {
  validates: Validate<A, I>[],
  extend(validate: Validate<A, I>): Validator<A, I>,
}

export class StringValidator extends t.Type<string> implements Validator<string, any> {
  validates = [
    (input: unknown, context: t.Context): Either<t.Errors, string> => 
      typeof input === 'string' ? t.success(input) : t.failure(input, context),
  ];

  constructor() {
    super(
      'string',
      (u): u is string => typeof u === 'string',
      (u, c) => {
        let e: Either<t.Errors, string> | undefined = undefined;
        for (const validate of this.validates) {
          e = validate(u, c);
          if (isLeft(e)) {
            return e;
          }
        }
        return e!;
      },
      t.identity
    )
  }

  extend(validate: Validate<string, any>) {
    const clone = new StringValidator();
    clone.validates = [...this.validates, validate];
    return clone;
  }

  refine(refiner: (s: string) => boolean, message?: string) {
    return this.extend(
      (input: string, context) => refiner(input)
        ? t.success(input)
        : t.failure(input, context, message),
    )
  }

  length(n: number, message?: string) {
    return this.refine((s) => s.length === n, message);
  }

  required(message?: string) {
    return this.min(1, message);
  }

  max(n: number, message?: string) {
    return this.refine((s) => s.length <= n, message);
  }

  min(n: number, message?: string) {
    return this.refine((s) => s.length >= n, message);
  }

  matches(regExp: RegExp, message?: string) {
    return this.refine((s) => regExp.test(s), message);
  }
}

export const string = () => new StringValidator();