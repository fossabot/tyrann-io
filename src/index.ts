export * from './types/basicTypes';
export * from './types/combinationTypes';
export * from './types/Either';
export * from './types/ErrorMessage';
export * from './types/infer';
export * from './types/LiteralType';
export * from './types/Resolver';
export * from './types/StructType';
export * from './types/utils';
