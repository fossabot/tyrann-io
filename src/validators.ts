import { Either, isLeft, isRight } from 'fp-ts/lib/Either';
import * as t from 'io-ts';
import { Encode } from 'io-ts';

export type Validate<A, I> = (input: I, context: t.Context) => Either<t.Errors, A>; 

export class Validator<A> extends t.Type<A> {
  validates: Validate<A, any>[] = [
    (input: unknown, context: t.Context): Either<t.Errors, A> => 
      this.is(input) ? t.success(input) : t.failure(input, context),
  ];

  constructor(
    name: string,
    is: t.Is<A>,
    encode: Encode<A, A>
  ) {
    super(
      name,
      is,
      (u, c) => {
        return this.validateAll(u, c);
      },
      encode,
    )
  }

  validateAll(u: unknown, c: t.Context) {
    let e: Either<t.Errors, A> | undefined = undefined;
    for (const validate of this.validates) {
      e = validate(u, c);
      if (isLeft(e)) {
        return e;
      } else {
        u = e.right;
      }
    }
    return e!;
  }

  clone(v: Validator<A>) {
    const c = new Validator(
      v.name,
      v.is,
      v.encode,
    );
    c.validates = v.validates;
    return c;
  }
  
  extend(validate: Validate<A, any>): Validator<A> {
    const clone = this.clone(this);
    clone.validates = [...this.validates, validate];
    return clone;
  }

  refine(refiner: (s: A) => boolean, message?: string) {
    return this.extend(
      (input: A, context) => refiner(input)
        ? t.success(input)
        : t.failure(input, context, message),
    )
  }
}

export class StringValidator extends Validator<string> {
  constructor() {
    super(
      'string',
      (u): u is string => typeof u === 'string',
      t.identity
    )
  }

  clone(v: StringValidator) {
    const c = new StringValidator();
    c.validates = v.validates;
    return c;
  }

  refine(refiner: (s: string) => boolean, message?: string): StringValidator {
    return super.refine(refiner, message) as StringValidator;
  }

  length(n: number, message?: string) {
    return this.refine((s) => s.length === n, message);
  }

  required(message?: string) {
    return this.min(1, message);
  }

  max(n: number, message?: string) {
    return this.refine((s) => s.length <= n, message);
  }

  min(n: number, message?: string) {
    return this.refine((s) => s.length >= n, message);
  }

  matches(regExp: RegExp, message?: string) {
    return this.refine((s) => regExp.test(s), message);
  }
}

export class NumberValidator extends Validator<number> {
  constructor() {
    super(
      'number',
      (u): u is number => typeof u === 'number',
      t.identity
    )
  }

  refine(refiner: (s: number) => boolean, message?: string): NumberValidator {
    return super.refine(refiner, message) as NumberValidator;
  }

  clone(v: NumberValidator) {
    const c = new NumberValidator();
    c.validates = v.validates;
    return c;
  }

  // Clear all validates, casting anything castable
  cast(message?: string) {
    const c = this.clone(this);
    c.validates = [
      (input: unknown, context: t.Context): Either<t.Errors, number> => 
        !Number.isNaN(Number(input))
        ? t.success(Number(input))
        : t.failure(input, context, message),
    ];
    return c;
  }

  min(n: number, message?: string) {
    return this.refine((s) => s >= n, message);
  }

  max(n: number, message?: string) {
    return this.refine((s) => s <= n, message);
  }

  positive(message?: string) {
    return this.min(0, message);
  }

  negative(message?: string) {
    return this.max(0, message);
  }

  integer(message?: string) {
    return this.refine((s) => Number.isSafeInteger(s), message);
  }
}

export class ArrayValidator<C extends t.Mixed> extends Validator<t.TypeOf<C>[]> {
  arrayType: t.ArrayType<C>;
  itemType: C;

  constructor(item: C) {
    super(
      `array<${item.name}>`,
      (u): u is Array<t.TypeOf<C>> => this.arrayType.is(u),
      (u) => this.arrayType.encode(u),
    );

    this.itemType = item;
    this.arrayType = t.array(item);
    this.validates.push((u, c) => this.arrayType.validate(u, c));
  }

  clone(v: ArrayValidator<C>) {
    const c = new ArrayValidator<C>(v.itemType);
    c.arrayType = v.arrayType;
    c.itemType = v.itemType;
    c.validates = v.validates;
    return c;
  }

  refine(refiner: (s: t.TypeOf<C>[]) => boolean, message?: string): ArrayValidator<C> {
    return super.refine(refiner, message) as ArrayValidator<C>;
  }

  min(n: number, message?: string) {
    return this.refine((s) => s.length >= n, message);
  }

  max(n: number, message?: string) {
    return this.refine((s) => s.length <= n, message);
  }

  required(message?: string) {
    return this.min(1, message);
  }
}
