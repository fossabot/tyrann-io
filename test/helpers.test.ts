import * as t from 'io-ts'
import * as h from '../src/helpers';


describe('helpers', () => {
  it('can validate', () => {
    const minLengthString = h
        .string()
        .min(5, 'Too short. ');

    expect(minLengthString.decode('123456')).toMatchSnapshot();
    expect(minLengthString.decode('1234')).toMatchSnapshot();

    const maxLengthString = h
        .string()
        .max(5, 'Too long. ');

    expect(maxLengthString.decode('123456')).toMatchSnapshot();
    expect(maxLengthString.decode('1234')).toMatchSnapshot();
  });

});
