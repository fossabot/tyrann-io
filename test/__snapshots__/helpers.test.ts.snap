// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers can validate 1`] = `
Object {
  "_tag": "Right",
  "right": "123456",
}
`;

exports[`helpers can validate 2`] = `
Object {
  "_tag": "Left",
  "left": Array [
    Object {
      "context": Array [
        Object {
          "actual": "1234",
          "key": "",
          "type": StringValidator {
            "decode": [Function],
            "encode": [Function],
            "is": [Function],
            "name": "string",
            "validate": [Function],
            "validates": Array [
              [Function],
              [Function],
            ],
          },
        },
      ],
      "message": "Too short. ",
      "value": "1234",
    },
  ],
}
`;

exports[`helpers can validate 3`] = `
Object {
  "_tag": "Left",
  "left": Array [
    Object {
      "context": Array [
        Object {
          "actual": "123456",
          "key": "",
          "type": StringValidator {
            "decode": [Function],
            "encode": [Function],
            "is": [Function],
            "name": "string",
            "validate": [Function],
            "validates": Array [
              [Function],
              [Function],
            ],
          },
        },
      ],
      "message": "Too long. ",
      "value": "123456",
    },
  ],
}
`;

exports[`helpers can validate 4`] = `
Object {
  "_tag": "Right",
  "right": "1234",
}
`;
